base/exports.jl

```diff
@@ -475,6 +475,7 @@ export
     pop!,
     popat!,
     prepend!,
+    push,
```

Comment:
As I explained in https://github.com/JuliaLang/julia/pull/24836#issuecomment-643067749, it's better to avoid exporting this.

Reply:
I don't understand well this other comment in the context of this present PR, where the meaning of `push` is made minimal: `push(dict::ImmutableDict, pair...)` cannot mean anything else I think than what it does in this PR. It doesnt seem breaking to add widening so that another dictionary type is returned (or very slightly so, in case we widen the type instead of convert the new pair to the `eltype` of `dict` but `ImmutableDict`). In any case, I'm fine not exported it but I don't understand what this changes, i.e. I don't see how "It is a big loss if we lose this API" depends on whether the symbol is exported or not...  is it about making it public API (that has to be stable) or not?

Reply:
The loss I mentioned was about `push(::AbstractVector, _)` and `push` on mutable containers in general. The major concern in #24836 for implementing `push(::AbstractVector, _)` was that some new users might try `push` instead of `push!`.  Not exporting `push` allows us to add `push(::AbstractVector, _)` without worrying about confusing new users.  (Implementing `push(::AbstractVector, _)` doesn't have to be done in this PR, of course.)

Reply:
Ok I understand now. I'm not conviced (we have other non-mutating functions which are exported), but the `export` can be the subject of another PR then.

Reply:
Yeah, I fully agree that it's not convincing that defining (e.g.) `merge`/`merge!` is OK and `push`/`push!` is not. But it's better to avoid the situation that we can't define `push` on `AbstractVector` just because it's exported.